lib/Tools/LinearLayout.cpp

```diff
@@ -804,118 +780,179 @@ LinearLayout LinearLayout::compose(const LinearLayout &outer) const {
     int32_t basis = 0;
-    for (int r = 0; r < numRowsOuter; r++) {
-      basis |= (m[r] >> (numColsOuter + c) & 1) << r;
+    for (int r = 0; r < numColsA; r++) {
+      basis |= (retMat[r] >> c & 1) << r;
     }
     bs.push_back({basis});
   }
 
-  LinearLayout flatComposed(std::move(newBases),
-                            {{outDim1D, outer.getTotalInDimSize()}},
+  LinearLayout retFlattened(std::move(retBases),
+                            {{outDim1D, A.getTotalInDimSize()}},
                             /*requireSurjective=*/false);
 
   SmallVector<std::pair<StringAttr, int32_t>> retInDims;
   SmallVector<std::pair<StringAttr, int32_t>> retOutDims;
-  for (StringAttr dim : getInDimNames()) {
-    retInDims.push_back({dim, getInDimSize(dim)});
+  for (StringAttr dim : B.getInDimNames()) {
+    retInDims.push_back({dim, B.getInDimSize(dim)});
   }
-  for (StringAttr dim : outer.getInDimNames()) {
-    retOutDims.push_back({dim, outer.getInDimSize(dim)});
+  for (StringAttr dim : A.getInDimNames()) {
+    retOutDims.push_back({dim, A.getInDimSize(dim)});
   }
-  return flatComposed.reshapeIns(retInDims).reshapeOuts(retOutDims);
+  return retFlattened.reshapeIns(retInDims).reshapeOuts(retOutDims);
+}
+
+} // namespace
+
+LinearLayout LinearLayout::invertAndCompose(const LinearLayout &outer) const {
+  // TODO(Lezcano) Make friend and perhaps rename to `convertFrom`
+  // For this, we need to implement our LLVM lowerings by inverting the "from"
+  // layout, and then iterating over the elements from the "to" layout and
+  // fetching the corresponding element from the "from" layout. This exercises
+  // the broadcasting that we incentivise via choosing the minimum norm solution
+  // in lstsq.
+
+  // The order of dims does not matter. We choose to transpose outer
+  auto outDims = llvm::to_vector(getOutDimNames());
+  assertDimsEqualIgnoringOrder(outDims, outer.getOutDimNames());
+  const auto &B = *this;
+  const auto A = outer.transposeOuts(outDims);
+  for (auto dim : outDims) {
+    assert(A.getOutDimSize(dim) == B.getOutDimSize(dim) &&
+           "Convert layout does not change the shape of a tensor");
+  }
+
+  // We'll write A^{-1} to mean the inverse or the pseudo-inverse of A
+  // We are computing A^{-1}B so from must be surjective so that
+  // it has a left inverse.
+  assert(A.isSurjective());
+
+  // Broadcasting heuristic
+  // Imagine we have two layouts with `warps = [[0, 0],  [0, 0]]`
+  // (broadcasting) on both layouts. We could map any warp to any warp in the
+  // conversion. Now, we want to map them as the identity map, to mark that
+  // nothing needs to be done there (`lstsq` would map all the warps to the
+  // zero warp, minimum norm solution). The heuristic here is as follows:
+  // - If a dimension is the same for both layouts, we want to map it as the
+  // identity
+  //   Equivalently, we don't add it to the conversion
+  // - Otherwise, we just call lstsq (i.e. map all the equivalent elements
```

Comment:
Not really. A^T is a different kind of object (albeit of the same shape). Here you need in general what's called the [pseudo-inverse](https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse.). See If you've never come across this object, the best description of what's going on in the real case (in terms of the SVD) is in [this section](https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse#Singular_value_decomposition_(SVD)).

That being said, you don't need to know about these objects really. Here's an explanation in terms of systems of linear equations:

What's happening here is that A is a wide matrix, so we want to solve `AX = B` where `A` is `mxk` with `k >= m`, `B` is `mxn` and `X = kxn`. Then, since A is surjective, we know that it has m independent columns. To avoid tracking indices, let's assume they are the first `m` (but they could be any subset of size m of the k columns it has). This means that we can split `A`  as `A = [A_1, A_2]` where `A_1` is `mxm` and invertible, and `A_2` is `mx(k-m)`. So we now write the system as:
```
[ A_1  A_2 ] [ X_1 ] = A_1 X_1 + A_2 X_2 = B
             [ X_2 ]
```
then, since `A_1` is invertible, we multiply everything  by A_1^{-1} to get:
```
X_1 = A^{-1}B - A^{-1}A_2X_2
```
where X_2 is any matrix of our choice. In this case, we choose `X_2 = 0` because of the broadcasting considerations, so we get our solution:
```
X = [ A^{-1} B]
    [   0     ]
```
This is exactly what's going on here.

Reply:
Got it now. Thanks for the explanation. Can you add [pseudo-inverse](https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse.) into the code comments?

Reply:
I plan to do a proper write-up, but in the meantime I can add a link to this comment in the code.